tools/common/Driver.cs

```diff
@@ -809,5 +787,162 @@ internal static bool ParseBool (string value, string name, bool show_error = tru
 			return result;
 		}
 
+		static readonly Dictionary<string, string> tools = new Dictionary<string, string> ();
+		static string FindTool (string tool)
+		{
+			string path;
+
+			lock (tools) {
+				if (tools.TryGetValue (tool, out path))
+					return path;
+			}
+
+			path = LocateTool (tool);
+			static string LocateTool (string tool)
+			{
+				if (XcrunFind (tool, out var path))
+					return path;
+
+				// either /Developer (Xcode 4.2 and earlier), /Applications/Xcode.app/Contents/Developer (Xcode 4.3) or user override
+				path = Path.Combine (DeveloperDirectory, "usr", "bin", tool);
+				if (File.Exists (path))
+					return path;
+
+				// Xcode 4.3 (without command-line tools) also has a copy of 'strip'
+				path = Path.Combine (DeveloperDirectory, "Toolchains", "XcodeDefault.xctoolchain", "usr", "bin", tool);
+				if (File.Exists (path))
+					return path;
+
+				// Xcode "Command-Line Tools" install a copy in /usr/bin (and it can be there afterward)
+				path = Path.Combine ("/usr", "bin", tool);
+				if (File.Exists (path))
+					return path;
+
+				return null;
+			}
+
+			// We can end up finding the same tool multiple times.
+			// That's not a problem.
+			lock (tools)
+				tools [tool] = path;
+
+			if (path == null)
+				throw ErrorHelper.CreateError (5307, Errors.MX5307 /* Missing '{0}' tool. Please install Xcode 'Command-Line Tools' component */, tool);
+
+			return path;
+		}
+
+		static bool XcrunFind (string tool, out string path)
+		{
+			return XcrunFind (ApplePlatform.None, false, tool, out path);
+		}
+
+		static bool XcrunFind (ApplePlatform platform, bool is_simulator, string tool, out string path)
+		{
+			var env = new List<string> ();
+			// Unset XCODE_DEVELOPER_DIR_PATH. See https://github.com/xamarin/xamarin-macios/issues/3931.
+			env.Add ("XCODE_DEVELOPER_DIR_PATH");
+			env.Add (null);
+			// Set DEVELOPER_DIR if we have it
+			if (!string.IsNullOrEmpty (DeveloperDirectory)) {
+				env.Add ("DEVELOPER_DIR");
+				env.Add (DeveloperDirectory);
+			}
```

Comment:
^ does the above change ? or can it be cached and reused

Reply:
You mean cache the list of environment values? Not sure if that's necessary, because we cache the result from this function.

Reply:
It's more for consistency across invocations than time savings. If we believe it can change then the caching is not really correct. If it can't change then it does not need to be repeated for every tool being queried. Current code makes it ambiguous (but it's not a big thing)